invenio_rdm_migrator/state.py

```diff
@@ -0,0 +1,248 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2023 CERN.
+#
+# Invenio-RDM-Migrator is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""Base state module."""
+
+from abc import ABC, abstractclassmethod
+from pathlib import Path
+from uuid import UUID
+
+from sqlalchemy import (
+    Column,
+    Integer,
+    MetaData,
+    String,
+    Table,
+    create_engine,
+    insert,
+    select,
+    update,
+)
+from sqlalchemy_utils.types import JSONType, UUIDType
+
+from .logging import Logger
+from .utils import ts
+
+
+class State:
+    """Migration state."""
+
+    def __init__(self, dbpath, validators=None):
+        """Constructor.
+
+        :param dbpath: path to sqlite db file.
+        """
+        # for extra validation that cannot be achieved with SQL constraints
+        self.validators = validators or {}
+        self.dbpath = Path(dbpath) / "state.db"
+        self.mem_eng = create_engine(f"sqlite:///:memory:")
+        self.metadata = MetaData()
+        self._initialize_db(self.metadata)  # bind tables to metadata
+
+        start = ts(iso=False)
+        logger = Logger.get_logger()
+        logger.info(f"Loading state from {self.dbpath} {ts()}")
+
+        disk_eng = create_engine(f"sqlite:///{self.dbpath}")
+        self.metadata.create_all(disk_eng)  # create tables if not exist
+
+        self._copy_db(disk_eng, self.mem_eng)  # dump state into in-memory sqlite db
+
+        end = ts(iso=False)
+        logger.info(f"Finished loading state {ts()}")
+        logger.info(f"State loading took {end-start} seconds.")
```

Comment:
I do agree, nothing against the change. However, it will always be the case that we want to load from disk (at least with the current use cases). Therefore, this change translates to two lines of code in every place (mostly tests though).

Reply:
Yes, the explicit call is what I meant. Or even better/cleaner, you can make the initialized parts lazy as properties to avoid duplication.